src/diepvries/template_sql/effectivity_satellite_dml.sql

```diff
@@ -5,36 +5,42 @@
 -- current timestamp minus 4 hours. The four hours are subtracted as a "safety net" to avoid the insertion of
 -- duplicate records when the first version of a given driving key is being loaded by two processes running in parallel.
 -- This is unlikely to happen, but still better to play it on the safe side.
-SET min_timestamp = (
-                      SELECT
-                        COALESCE(MIN(satellite.{record_start_timestamp}), DATEADD(HOUR, -4, CURRENT_TIMESTAMP()))
-                      FROM {target_schema}.{link_table} AS l
-                        INNER JOIN {target_schema}.{target_table} AS satellite
-                                   ON (l.{hashkey_field} = satellite.{hashkey_field}
-                                      AND satellite.{record_end_timestamp_name} = {end_of_time})
-                        INNER JOIN {staging_schema}.{staging_table} AS staging
-                                   ON ({link_driving_key_condition})
-                      );
+SET min_timestamp_link = (
+                         SELECT
+                           COALESCE(MIN(l.{record_start_timestamp}), DATEADD(HOUR, -4, CURRENT_TIMESTAMP()))
+                         FROM {target_schema}.{link_table} AS l
+                           INNER JOIN {staging_schema}.{staging_table} AS staging
+                                      ON ({link_driving_key_condition})
+                         );
+
+SET min_timestamp_satellite = (
+                              SELECT
+                                COALESCE(MIN(satellite.{record_start_timestamp}),
+                                         DATEADD(HOUR, -4, CURRENT_TIMESTAMP()))
+                              FROM {target_schema}.{link_table} AS l
+                                INNER JOIN {target_schema}.{target_table} AS satellite
+                                           ON (l.{hashkey_field} = satellite.{hashkey_field}
+                                             AND satellite.{record_end_timestamp_name} = {end_of_time}
+                                             AND l.{record_start_timestamp} >= $min_timestamp_link)
+                                INNER JOIN {staging_schema}.{staging_table} AS staging
+                                           ON ({link_driving_key_condition})
+                              );
 
 MERGE INTO {target_schema}.{target_table} AS satellite
   USING (
         WITH
-          effectivity_satellite AS (
+          filtered_effectivity_satellite AS (
           SELECT
             {link_driving_keys},
             satellite.*
-          FROM {target_schema}.{link_table} AS l
+          FROM {staging_schema}.{staging_table} AS staging
+            INNER JOIN {target_schema}.{link_table} AS l
+                       ON ({link_driving_key_condition}
+                         AND l.{record_start_timestamp} >= $min_timestamp_link)
```

Comment:
Logically speaking I'd think the same, but the evidence says otherwise unfortunately. It seems that for it to be super quick having filters in both tables is the best way. Even though we end up running three queries instead of one, the number of bytes scanned and overall execution time is better. Let me gather some query ids from my tests

Reply:
Hmmmn I guess it would make sense that filtering on a timestamp is easier than filtering through a `JOIN` on keys.

Reply:
Indeed and that's exactly what happens, but the fact is that by calculating the `min_timestamp_link` and `min_timestamp_satellite` we're also doing joins between tables without specific filters. The fact that the calculation of these variables is way faster is because the boundaries of each timestamp `MIN/MAX` are stored in the table's statistics hence snowflake can fetch them quicker (my theory!)